pkg/federate/kubefed/resource.go

```diff
@@ -1,17 +1,114 @@
+	errorUpdateResource         = "Error updating the resource, assuming it does not exist and creating new one: %v"
+	errorCreateResource         = "Error creating the resource: %v"
 )
 
+var systemMetadataFields = []string{"selfLink", "uid", "resourceVersion", "generation", "creationTimestamp", "deletionTimestamp", "deletionGracePeriodSeconds"}
+
 func (f *federator) Distribute(resource runtime.Object, clusterNames ...string) error {
-	panic("not implemented")
+	fedResource, err := f.createFederatedResource(resource)
+	if err != nil {
+		return err
+	}
+
+	if len(clusterNames) > 0 {
+		unstructured.RemoveNestedField(fedResource.Object, ctlutil.SpecField, ctlutil.PlacementField)
+		err = ctlutil.SetClusterNames(fedResource, clusterNames)
+		if err != nil {
+			// Path not covered by unit tests because it's trivial behavior yet complex to test
+			return fmt.Errorf(errorSettingFederatedFields, err)
+		}
+	}
+
+	// Update first, as it will probably be the most common operation
+	err = f.kubeFedClient.Update(context.TODO(), fedResource)
+	if err == nil {
+		return nil
+	}
+
+	klog.Infof(errorUpdateResource, err)
+
+	err = f.kubeFedClient.Create(context.TODO(), fedResource)
+	if err != nil {
+		// Path not covered by unit tests because it's trivial behavior yet complex to test
+		return fmt.Errorf(errorCreateResource, err)
+	}
+
+	return nil
 }
 
 func (f *federator) Delete(resource runtime.Object) error {
 	panic("not implemented")
 }
+
+func (f *federator) createFederatedResource(resource runtime.Object) (*unstructured.Unstructured, error) {
+	targetResource := &unstructured.Unstructured{}
+	err := f.scheme.Convert(resource, targetResource, nil)
+	if err != nil {
+		return nil, fmt.Errorf(errorUnstructuredConversion, err)
+	}
+
+	fedResource := &unstructured.Unstructured{}
+	setBasicMetaFields(fedResource, targetResource)
+
+	// TODO(mpeterson): Figure out how to use the one from KubeFed
+	// kubefedfed.RemoveUnwantedFields(targetResource)
+	removeUnwantedFields(targetResource)
+
+	err = unstructured.SetNestedField(fedResource.Object, targetResource.Object, ctlutil.SpecField, ctlutil.TemplateField)
+	if err != nil {
+		// Path not covered by unit tests because it's trivial behavior yet complex to test
+		return nil, fmt.Errorf(errorSettingFederatedFields, err)
+	}
+	err = unstructured.SetNestedStringMap(fedResource.Object, map[string]string{}, ctlutil.SpecField, ctlutil.PlacementField, ctlutil.ClusterSelectorField, ctlutil.MatchLabelsField)
+	if err != nil {
+		// Path not covered by unit tests because it's trivial behavior yet complex to test
+		return nil, fmt.Errorf(errorSettingFederatedFields, err)
+	}
+
+	return fedResource, nil
+}
+
+func setBasicMetaFields(resource, base *unstructured.Unstructured) {
+	fedKind := federatedKindPrefix + base.GetKind()
+	resource.SetKind(fedKind)
+
+	gv := schema.GroupVersion{
+		Group:   kubefedopt.DefaultFederatedGroup,
+		Version: kubefedopt.DefaultFederatedVersion,
```

Comment:
Yeah, actually the most correct way to do this is to receive the TypeConfig just in case the user defined other API versions that are not the default (it is not going to be our case that's why it makes a lot of sense to use the defaults here)